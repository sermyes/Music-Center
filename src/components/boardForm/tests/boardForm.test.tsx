import BoardForm from '../boardForm';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { fakePosts } from './../../tests/posts';
jest.mock('uuid', () => {
  return {
    v4: () => '0'
  };
});

describe('Board Item', () => {
  const admin = fakePosts.admin!;
  const onConfirm = jest.fn();
  let option: HTMLSelectElement;
  let id: HTMLInputElement;
  let psw: HTMLInputElement;
  let content: HTMLTextAreaElement;
  let button: HTMLButtonElement;
  window.alert = jest.fn();
  it('renders correctly', () => {
    const component = renderer.create(
      <BoardForm onConfirm={onConfirm} admin={admin} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  beforeEach(() => {
    render(<BoardForm onConfirm={onConfirm} admin={admin} />);
    option = screen.getByTestId('option');
    id = screen.getByTestId('id');
    psw = screen.getByTestId('psw');
    content = screen.getByTestId('content');
    button = screen.getByRole('button');
  });

  describe('enter notice with admin', () => {
    it('if you have administrator rights', async () => {
      await userEvent.selectOptions(option, ['notice']);
      await userEvent.type(id, 'admin');
      await userEvent.type(psw, '123');
      await userEvent.type(content, 'content');
      await userEvent.click(button);

      expect(onConfirm).toHaveBeenCalledTimes(1);
      expect(onConfirm.mock.calls[0][1]).toBe('0');
      expect(onConfirm.mock.calls[0][0]['0']['name']).toBe('admin');
      expect(onConfirm.mock.calls[0][0]['0']['psw']).toBe('123');
      expect(onConfirm.mock.calls[0][0]['0']['content']).toBe('content');
      expect(onConfirm.mock.calls[0][0]['0']['type']).toBe('notice');
    });

    it('if you do not have administrator rights', async () => {
      await userEvent.selectOptions(option, ['notice']);
      await userEvent.type(id, 'admin');
      await userEvent.type(psw, '1');
      await userEvent.type(content, 'content');
      await userEvent.click(button);

      expect(onConfirm).toHaveBeenCalledTimes(0);
      expect(window.alert).toHaveBeenCalledTimes(1);
    });

    it('if author could not be entered', async () => {
      await userEvent.selectOptions(option, ['notice']);
      await userEvent.type(psw, '123');
      await userEvent.type(content, 'content');
      await userEvent.click(button);

      expect(onConfirm).toHaveBeenCalledTimes(0);
      expect(window.alert).toHaveBeenCalledTimes(1);
    });

    it('if password could not be entered', async () => {
      await userEvent.selectOptions(option, ['notice']);
      await userEvent.type(id, 'admin');
      await userEvent.type(content, 'content');
      await userEvent.click(button);

      expect(onConfirm).toHaveBeenCalledTimes(0);
      expect(window.alert).toHaveBeenCalledTimes(1);
    });

    it('if content could not be entered', async () => {
      await userEvent.selectOptions(option, ['notice']);
      await userEvent.type(id, 'admin');
      await userEvent.type(psw, '123');
      await userEvent.click(button);

      expect(onConfirm).toHaveBeenCalledTimes(0);
      expect(window.alert).toHaveBeenCalledTimes(1);
    });
  });

  it('enter question data', async () => {
    await userEvent.selectOptions(option, ['post']);
    await userEvent.type(id, 'user');
    await userEvent.type(psw, '1');
    await userEvent.type(content, 'content');
    await userEvent.click(button);

    expect(onConfirm).toHaveBeenCalledTimes(1);
    expect(onConfirm.mock.calls[0][1]).toBe('0');
    expect(onConfirm.mock.calls[0][0]['0']['name']).toBe('user');
    expect(onConfirm.mock.calls[0][0]['0']['psw']).toBe('1');
    expect(onConfirm.mock.calls[0][0]['0']['content']).toBe('content');
    expect(onConfirm.mock.calls[0][0]['0']['type']).toBe('post');
  });

  it('enter question data', async () => {
    await userEvent.selectOptions(option, ['question']);
    await userEvent.type(id, 'user');
    await userEvent.type(psw, '1');
    await userEvent.type(content, 'content');
    await userEvent.click(button);

    expect(onConfirm).toHaveBeenCalledTimes(1);
    expect(onConfirm.mock.calls[0][1]).toBe('0');
    expect(onConfirm.mock.calls[0][0]['0']['name']).toBe('user');
    expect(onConfirm.mock.calls[0][0]['0']['psw']).toBe('1');
    expect(onConfirm.mock.calls[0][0]['0']['content']).toBe('content');
    expect(onConfirm.mock.calls[0][0]['0']['type']).toBe('question');
  });
});
