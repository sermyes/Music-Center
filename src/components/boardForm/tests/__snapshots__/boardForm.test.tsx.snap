// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board Item renders correctly 1`] = `
<form>
  <fieldset
    className="input"
  >
    <div
      className="typeContainer"
    >
      <select
        className="type"
        data-testid="option"
        defaultValue="post"
        id="type"
        name="type"
        onChange={[Function]}
      >
        <option
          className="notice"
          value="notice"
        >
          Notice
        </option>
        <option
          className="request"
          value="post"
        >
          Request
        </option>
        <option
          className="question"
          value="question"
        >
          Talk
        </option>
      </select>
    </div>
    <div
      className="info"
    >
      <label
        htmlFor="id"
      >
        <i
          className="fas fa-user"
        />
      </label>
      <input
        data-testid="id"
        id="id"
        maxLength={8}
        minLength={1}
        required={true}
        type="text"
      />
      <label
        htmlFor="psw"
      >
        <i
          className="fas fa-unlock"
        />
      </label>
      <input
        autoComplete="on"
        data-testid="psw"
        id="psw"
        maxLength={10}
        minLength={1}
        required={true}
        type="password"
      />
    </div>
    <div
      className="wrapper"
    >
      <textarea
        className="content"
        cols={30}
        data-testid="content"
        required={true}
        rows={10}
      />
      <button
        className="confirm"
        data-testid="confirm"
        onClick={[Function]}
      >
        Post Reply
      </button>
    </div>
  </fieldset>
</form>
`;
